File: golang/2303.calculate-amount-paid-in-taxes.go
func calculateTax(brackets [][]int, income int) float64 {
	tax := 0.0
	prev := 0

	for i := 0; i < len(brackets) && income > 0; i++ {
		base := min(income, brackets[i][0]-prev)

		tax += float64(base) * float64(brackets[i][1]) / 100.0
		income -= base

		prev = brackets[i][0]
	}

	return tax
}

File: python/2303.calculate-amount-paid-in-taxes.py
class Solution:
    def calculateTax(self, brackets: list[list[int]], income: int) -> float:

        prev, tax = 0, 0.0

        for bound, percentage in brackets:
            base = min(income, bound - prev)
            income -= base
            tax += base * percentage / 100

            prev = bound

            if income == 0:
                break

        return tax

